### PR TITLE
Subscriptions 相關 graphql

### LINKED ISSUE #762: [API] Query mySubcriptions

### PR DESCRIPTION
**Pull request by mark86092**
close #762, #765

```
src/models/schemas/   定義 mongoose schema，與 DB 儲存的格式有關
src/schemas/   定義 graphql schema，與 API 格式有關
```

`resolver` 將 DB 的格式與 graphql 格式作轉換

### CODE REVIEW

File: src/schema/subscription.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    SkuRecord: {
+        __resolveType() {
+            // resolve to graphql schema Subscription
+            return "Subscription";
+        },
+    },
+    Subscription: {
+        id: subscription => subscription._id,
+        start_at: subscription => subscription.created_at,
+        end_at: subscription => subscription.expired_at, // TODO, check
```
Comment (mark86092) [inline src/schema/subscription.js:54]: 從 API 文件，暫時得到 DB 是 expired_at, API 是 end_at

File: src/schema/subscription.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    },
+    Query: {
+        get_my_subcriptions: combineResolvers(
```
Comment (mark86092) [inline src/schema/subscription.js:57]: 待討論：snake_case or camelCase  ?

File: src/models/schemas/supscription_schema.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+supscription_schema.query.isValidAt = function(current_date) {
+    if (!current_date) {
+        current_date = new Date();
+    }
+    return this.where({
+        status: SUBSCRIPTION_STATUS.OK,
+        created_at: { $lte: current_date },
+        expired_at: { $gte: current_date },
+    });
```
Comment (mark86092) [inline src/models/schemas/supscription_schema.js:53]: 這邊用來定義，何謂有效的 subscription

File: src/models/schemas/supscription_schema.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+const supscription_schema = new Schema(
+    {
+        payment_record_id: {
+            type: mongoose.ObjectId,
+            index: true,
+        },
+        user_id: {
+            type: mongoose.ObjectId,
+            index: true,
```
Comment (mark86092) [inline src/models/schemas/supscription_schema.js:13]: required: true

File: src/models/schemas/subscription_schema.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    },
+    {
+        timestamps: {
+            createdAt: "created_at",
+            updatedAt: "updated_at",
+        },
+    }
+);
+
+subscription_schema.query.isValidAt = function(current_date) {
```
Comment (mark86092) [inline src/models/schemas/subscription_schema.js:46]: 這邊用來定義，何謂有效的 subscription

File: src/schema/subscription.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    interface SkuRecord {
+        id: ID!
+    }
+
+    type Subscription implements SkuRecord {
+        id: ID!
+        start_at: Date!
+        end_at: Date!
```
Comment (mark86092) [inline src/schema/subscription.js:25]: Change to 

```
    started_at: Date!
    expired_at: Date!
```

File: src/schema/subscription.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+const resolvers = {
+    SkuRecord: {
+        __resolveType() {
+            // resolve to graphql schema Subscription
+            return "Subscription";
+        },
+    },
+    Subscription: {
+        id: subscription => subscription._id,
+        start_at: subscription => subscription.created_at,
```
Comment (mark86092) [inline src/schema/subscription.js:53]: started_at

File: src/schema/subscription.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        end_at: subscription => subscription.expired_at, // TODO, check
```
Comment (mark86092) [inline src/schema/subscription.js:54]: expired_at

File: src/models/schemas/subscription_schema.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            type: Schema.Types.Mixed,
+        },
+        extra_data: {
+            experience_id: {
+                type: mongoose.ObjectId,
+            },
+            salary_id: {
+                type: mongoose.ObjectId,
+            },
+        },
```
Comment (mark86092) [inline src/models/schemas/subscription_schema.js:32]: 多 started_at

File: src/models/schemas/subscription_schema.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    }
+);
+
+subscription_schema.query.isValidAt = function(current_date) {
+    if (!current_date) {
+        current_date = new Date();
+    }
+    return this.where({
+        status: SUBSCRIPTION_STATUS.OK,
+        created_at: { $lte: current_date },
```
Comment (mark86092) [inline src/models/schemas/subscription_schema.js:52]: started_at

File: src/schema/subscription.test.js [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+                current_date.toISOString()
+            );
+            expect(subscriptions[1]).to.have.property(
+                "status",
+                SUBSCRIPTION_STATUS.INIT
+            );
+        });
+    });
+
+    describe("query.get_my_current_subscription", () => {
```
Comment (barry800414) [inline src/schema/subscription.test.js:123]: How about adding a test case for `no subscription at all` scenario ? 
Which may apply to many of our users.

**Review by barry800414 — APPROVED** [review]

File: src/schema/subscription.test.js (hunk 11 above)
Comment (mark86092) [reply src/schema/subscription.test.js:134]: I add test case for no subscription

https://github.com/goodjoblife/WorkTimeSurvey-backend/pull/778/commits/1d78fb89362666c1d97319d7c26e5a1dda95b1c1

**Pull request merged**